### ISSUE TITLE
WPF app not sending data

### ISSUE THREAD

**Issue by MikaelStalvik** [microsoft/appcenter-sdk-dotnet#1452]
### **Description**

I have included AppCenter in my WPF application.
I can see that both crashes and analytics traffic are properly sent by viewing my network traffic with Fiddler on one of my machines, my development machine.
This works fine in both debug and release mode.

When I switch to two other machines, I can no longer see any traffic through Fiddler.

My logs shows that both analytics and crashes are initialized.

### **Repro Steps**

1. Register analytics and crash reporting as in your sample.
2. View the outgoing traffic through Fiddler

Actual result: On two machines, no outgoing traffic is regitered, on one machine traffic is registered.
Expected result: Outgoing traffic shall be registered, and events shall be visible.

### **Details**

1. Platform: Win10, WPF. .NET 4.7.2.
2. SDK version 3.4.2.
3. OS: Windows 10
4. Third-party libraries: Telerik UI for WPF, DevArt SQLite connector, MySql, JSON .NET.
5. Logging works well when started from VisualStudio 2019 in release mode, but I cannot test it on my other devices since I do not start them from Visual Studio 

I'm quite perplexed, please give me any hints of what I can try.

**Comment by vvechkanov** [comment]
Hi @MikaelStalvik! 

I tried to reproduce this issue, but I did not succeed. Could you please try to provide logs from applications, launched on other devices (where events not working)? To help you, we need to understand what is going wrong on machines where AppCenter is not working.

Also, are machines where  AppCenter is not working also Windows 10? Or maybe they use some specific OS or hardware.

**Comment by MikaelStalvik** [comment]
Thanks for your reply @vvechkanov.

>>To help you, we need to understand what is going wrong on machines where AppCenter is not working.
Is it possible to redirect AppCenters verbose logging from console only to a file-based logging?
At the moment I cannot see the logs on my devices where it does not work, since the logging is only presented in the console. which is not available.

>>Also, are machines where AppCenter is not working also Windows 10? Or maybe they use some specific OS or hardware.
All machines are running on the most recent Windows 10 version.
Can integrity settings in Windows 10 affect this?

**Comment by MikaelStalvik** [comment]
@vvechkanov :
I managed to extract the following logs from one of the machines where it does not work:

2020-09-29 19:23:10.604 [AppCenter] DEBUG: Invoke EnqueuingLog event for channel 'analytics'
2020-09-29 19:23:10.606 [AppCenter] DEBUG: Invoke FilteringLog event for channel 'analytics'
2020-09-29 19:23:10.607 [AppCenter] ERROR: Error persisting log
Microsoft.AppCenter.Storage.StorageException: The database wasn't initialized.
   at Microsoft.AppCenter.Storage.Storage.<>c__DisplayClass30_0.<AddTaskToQueue>b__0()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AppCenter.Channel.Channel.<PersistLogAsync>d__40.MoveNext()

What's happening behind the scene here?
The required SQLitePCL dlls are included (verified) although we use Devart.Data.SQLite.dll which requires a reference to sqlite3.dll.

Can this be the reason? If so, can we disable SQLite for AppCenter to get it to work?

**Comment by vvechkanov** [comment]
Hi!
Based on logs we see that database was not initialized correctly. Could you please check where AppCenter store database on your machine (the one where everything is not working). For it you need to execute this function:
`
public string getAppCenterDatabasePath() {
            var localAppData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            var appCenterFilesDirectoryPathBacking = System.IO.Path.Combine(localAppData, "Microsoft", "AppCenter", "installIdGoesHere");
            return System.IO.Path.Combine(appCenterFilesDirectoryPathBacking, "Logs.db");
        }
`
This function will return you a string. You need to check whether your application has write access to this directory.

**Comment by MikaelStalvik** [comment]
Hi!

I can see that this path resolves to: C:\Users\MY_USER_NAME\AppData\Local\Microsoft\AppCenter\fded1204-deb0-411d-b4d6-78285efdf19c

On my machine where this works, I can see that logs.db is properly created whilst only the folder is created on the machines where it does not work.

I use Windows with standard permissions and can create documents in this folder from Explorer/commandline without running any elevated processes.

Is it possible to configure the path to the database somehow?
I looked in the code but did not found any suitable method for this.

**Comment by vvechkanov** [comment]
HI! 

No, you could not change the path to the database, without changing AppCenter code. 
To check permissions, could you try to launch your app in the "administrator" mode? Will it work in this case?

**Comment by MikaelStalvik** [comment]
Thanks, unfortunately that gave the exact same output:

>
2020-10-05 17:16:11.007 [AppCenter] DEBUG: Invoke EnqueuingLog event for channel 'analytics'
2020-10-05 17:16:11.008 [AppCenter] DEBUG: Invoke FilteringLog event for channel 'analytics'
2020-10-05 17:16:11.009 [AppCenter] ERROR: Error persisting log
Microsoft.AppCenter.Storage.StorageException: The database wasn't initialized.
   at Microsoft.AppCenter.Storage.Storage.<>c__DisplayClass30_0.<AddTaskToQueue>b__0()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AppCenter.Channel.Channel.<PersistLogAsync>d__40.MoveNext()

**Comment by vvechkanov** [comment]
Hello!

Looks like the problem is not in the restrictions. Could you, please create a simple application with AppCenter (I mean just create a new app and add AppCenter into it) and check, will Crashes and analytics be reported from the problem machine.

**Comment by MikaelStalvik** [comment]
Sorry, I will need to rever to use ExceptionLess instead, I cannot spend any more time to investigate this at the moment.
I will try to migrate to AppCenter in the future again, when my time allows it.

**Comment by Nickjgniklu** [comment]
I am also seeing this error here is an error message that I have received. 
``System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> Microsoft.AppCenter.Storage.StorageException: The database wasn't initialized.
   at Microsoft.AppCenter.Storage.Storage.<>c__DisplayClass32_0`1.<AddTaskToQueue>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AppCenter.Storage.Storage.<FlushQueueAsync>d__35.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) Microsoft.AppCenter.Storage.StorageException: The database wasn't initialized.
   at Microsoft.AppCenter.Storage.Storage.<>c__DisplayClass32_0`1.<AddTaskToQueue>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AppCenter.Storage.Storage.<FlushQueueAsync>d__35.MoveNext()<---``

**Comment by Jamminroot** [comment]
Hey @Nickjgniklu, can you elaborate a bit:
What's the SDK version you are using?
Can the problem be reproduced on a clean project?
Is there a chance to have a look at verbose logs?
What are the nugets you are using (if you are able to disclose that type of information)?

Reopening the issue for now.

**Comment by Nickjgniklu** [comment]
Hi @Jamminroot! Thanks for getting back to me. I did some digging into this problem to try get the info you want and found the problem. I was only seeing the issue when our program was installed by our installer (using inno); it did not occur in a clean project or in the same project when it was ran from visual studio. I looked around to try and figure out why and noticed that the installer did not bring all the DLLs sqlite needed. In particular the bin folder generated by visual studio contained a runtimes folder which had some important sqlite dlls in it. 

This is what the file structure looks like when it is correct (generated by visual studio on build)
Project Folder
-bin
--Multiple dlls, exe and config files at this level
--runtimes (The misconfigured installer generator missed this whole folder of dlls)
---win-arm
----native
-----e_sqlite3.dll
---win-x64
----native
-----e_sqlite3.dll
---win-x86
----native
-----e_sqlite3.dll

This is what the file structure looks like when it is wrong (generated by our installer on install)
Project Folder
-bin
--Multiple dlls, exe and config files at this level

I don't think this is a bug but hopefully this can help anyone else who misconfigures their installer creator.
